### PR TITLE
CA TreeView

### PR DESCRIPTION
**Pull request by norisalsaadie**
Migrating the ca-treeview web component to the common-UI project.
This required migrating Ajax and secure Ajax, as well as other components.
Files were updated against new linting rules, and where possible, logic change was avoided

Changes:
.eslintrc -> adding ajax as global until a better solution is found.
components/ca-treeview.css -> tree-view css from old private repo
components/ca-treeview.js -> tree-view from old private repo, migrated to ES6.
components/helpers/ajax.js -> moved and cleaned a little, no AMD module yet.
components/helpers/secureajax.js -> moved and cleaned a little, no AMD module yet.
components/helpers/cancel-event.js -> migrate from old helpers.js.
components/helpers/clear-element.js -> migrated from old helpers.js.
components/helpers/parent-by-attribute.js -> migrated from old helpers.js.
examples/ca-treeview.html -> moved example, had to use promises due to AMD.
test/ca-tree-view-test.html -> migrated the old tests and refactored.
test/helpers/cancel-event-test.html -> placeholder tests with TODO.
test/helpers/clear-element-test.html -> placeholder tests with TODO.
test/helpers/parent-by-attribute-test.html -> placeholder tests with TODO.

### CODE REVIEW

File: components/helpers/ajax.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
+// generic ajax request function
```
Comment (kevmarchant) [inline components/helpers/ajax.js]: worth commenting that this needs refactoring into dependency rather than window.ajax?

Comment (norisalsaadie) [inline components/helpers/ajax.js]: Yes. I have already done so in another branch :)

File: components/ca-treeview.js [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+(createElement, parentByAttribute, cancelEvent) => {
+    /**
+     * Class TreeView for a TreeView web component
+     * @extends HTMLElement
+     */
+    class TreeView extends HTMLElement {
+        /**
+         * createdCallback is called when the component is first created.
+         * creates children, binds events and loads data if src is avaliable.
+         * @returns {undefined} nothing.
```
Comment (norisalsaadie) [inline components/ca-treeview.js]: Yes and no. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void

File: components/ca-treeview.css [hunk 3]
```diff
@@ -0,0 +1,34 @@
+/* GENERIC WEB-COMPONENT OVERRIDES */
+ca-treeview { display: inline-block; vertical-align: top; border-right: none !important; background-color: #E5E8EC; margin: 0; padding: 0.5em; position: absolute; box-sizing: border-box; width: 200px; }
```
Comment (norisalsaadie) [inline components/ca-treeview.css]: Yep, my view as well. But this is a copy pasta. I'm less fussed about the CSS initially.

File: components/ca-treeview.js (hunk 2 above)
Comment (tapina) [inline components/ca-treeview.js]: Actually it (or at least the type) [should be omitted altogether](http://stackoverflow.com/a/20650023)

File: components/ca-treeview.js [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+         */
+        get root() {
+            return this.getAttribute('data-root') || '';
+        }
+
+        /**
+         * Gets the root attribute
+         * @param {string} value, data-root attribute, default is empty.
+         */
+        set root(value = '') {
```
Comment (tapina) [inline components/ca-treeview.js]: This default value is not equivalent to the previous version. It's not even just a refactoring, it's a behaviour change. It used to say:
`                this.setAttribute('data-root', value || '');`
which would assign an empty string for any falsy value. In particular, passing null will behave differently.

File: components/helpers/ajax.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+     * @param {string} href to target
+     * @returns {string} encoded data
+     */
+    function parseDataUri(href) {
+        const commaPos = href.indexOf(',');
+        // header is of the form data:[<mediatype>][;base64]
+        const header = href.substring(0, commaPos);
+        let data = href.substring(commaPos + 1);
+        if (header.endsWith(';base64')) {
+            data = escape(window.atob(data));
```
Comment (tapina) [inline components/helpers/ajax.js]: NB this requires a polyfill in some browsers.

File: components/helpers/parent-by-attribute.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+define('parent-by-attribute', [], () =>
+    function getParentByAttribute(el, attName, attValue) {
```
Comment (tapina) [inline components/helpers/parent-by-attribute.js]: JSDoc? Blimey this is a monstrosity though. 

Add a TODO: replace with [Element.closest()](http://caniuse.com/#search=closest) and its [polyfill](https://github.com/jonathantneal/closest) for IE.

File: examples/ca-treeview.html [hunk 7]
```diff
@@ -0,0 +1,41 @@
+<!doctype html>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <title>CA NOTIFICATION CONTAINER | Web Lab Common UI</title>
```
Comment (tapina) [inline examples/ca-treeview.html]: It's not a notification container.

File: test/ca-tree-view-test.html [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                        </li>
+                      </ol>
+                    </nav>
+                  </ca-treeview>`;
+
+                // Force a render
+                element.render(element.ol, data);
+
+                // And check that the generated HTML matches
+                expect(element.outerHTML).to.equal(html.replace(/\s\s+/g, ''));
```
Comment (tapina) [inline test/ca-tree-view-test.html]: This makes all the rest of the checks pointless. But that's not your fault.

File: components/ca-treeview.css (hunk 3 above)
Comment (norisalsaadie) [inline components/ca-treeview.css]: We should.

File: components/ca-treeview.js (hunk 2 above)
Comment (norisalsaadie) [inline components/ca-treeview.js]: Yeah, for some reason, the linter errors if I omit the `@returns` - I don't like having a @returns if there is no return, it's very annoying.

File: components/ca-treeview.js [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+            ajax
+                .execute({ url: dataSource, dataType: 'json' })
+                .then(data => {
+                    // if this is a schema with instance links, grab the instance url
+                    const instancesUrl = (data.$schema && data.items.links) ? (data.items.links.find(link => link.rel === 'instances') || {}).href : '';
+                    if (instancesUrl !== '') {
+                        // we need to go get the instance data and recall this function to trigger rendering
+                        this.loadData(parent, instancesUrl);
+                    } else {
+                        createElement(this.ol);
```
Comment (norisalsaadie) [inline components/ca-treeview.js]: Good find. updated.

File: components/helpers/ajax.js (hunk 5 above)
Comment (norisalsaadie) [inline components/helpers/ajax.js]: Oh, that sucks.

File: test/ca-tree-view-test.html (hunk 8 above)
Comment (norisalsaadie) [inline test/ca-tree-view-test.html]: yep, copy-pasta + top and tail.

File: components/ca-treeview.js [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+         */
+        get maxChildren() {
+            return parseInt(this.getAttribute('data-max-children') || '0', 10);
+        }
+
+        /**
+         * Sets the max children allowed.
+         * @param {int} value number of allow childre, defaults to 0.
+         */
+        set maxChildren(value = 0) {
```
Comment (norisalsaadie) [inline components/ca-treeview.js]: But this **shouldn't** break by any means. parseInt(0, 10) -> 0.

File: components/ca-treeview.js [hunk 11]
```diff
@@ -0,0 +1,263 @@
+define(
+['/components/helpers/create-element.js', '/components/helpers/parent-by-attribute.js', '/components/helpers/cancel-event.js', 'document-register-element'],
```
Comment (norisalsaadie) [inline components/ca-treeview.js]: Done

File: components/ca-treeview.js [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+         * @param {string} tagName, the html tag to create.
+         * @param {object} item, schema object.
+         * @returns {HTMLElement} the newly created list item.
+         */
+        createChild(parent, tagName, item) {
+            const label = item.label || item.title || 'undefined';
+            const rel = (item.rel || '').toLowerCase().replace(/(describedby|described_by):/gi, '');
+
+            // get a list of attributes to be added to the element
+            const attributes = Object.keys(item).reduce((attributeList, key) => {
```
Comment (norisalsaadie) [inline components/ca-treeview.js]: You are right we can guard against it. But we adopted the airbnb rules as our base, and they ban for in https://github.com/airbnb/javascript#iterators-and-generators, and I'm not really for overriding the rules so easily. 

The use of a generic reduce was to maintain the loop's logic but get it passing linting. I agree with your suggestion of using chained filter and map, but I see this as part of a v1 refactor. What do you think?

File: components/ca-treeview.js (hunk 4 above)
Comment (norisalsaadie) [inline components/ca-treeview.js]: Good point, my bad. Reverting.

File: components/helpers/parent-by-attribute.js (hunk 6 above)
Comment (norisalsaadie) [inline components/helpers/parent-by-attribute.js]: I forgot to copy the JSdoc from the helpers.js. I have done so now, and included the TODO.

File: components/ca-treeview.js (hunk 10 above)
Comment (norisalsaadie) [inline components/ca-treeview.js]: Also, document.querySelector('body').setAttribute('baaa-baaa-moewww', 5) -> would produce <body baaa-baaa-moewww="5>..</body>

And fetching it document.querySelector('body').getAttribute('baaa-baaa-moewww') -> "5"

File: components/ca-treeview.js (hunk 2 above)
Comment (tapina) [inline components/ca-treeview.js]: Can you omit the type so it's `@returns nothing.` ? Or does it still complain?

Comment (norisalsaadie) [inline components/ca-treeview.js]: I recall trying this in older versions of eslint, and it complained. But I just tried this with our eslint version, and it seems fine - so let's do that :)

File: components/ca-treeview.js (hunk 12 above)
Comment (tapina) [inline components/ca-treeview.js]: I'm not sure I want to ban iterators? Are we just banning them in common-ui or did we ban them in all our linted projects? Airbnb claim on that page it's for immutability and avoiding side effects then promptly demonstrate a refactoring to a forEach loop with side effects just the same as the iterator it replaces.

Anyway, the generic reduce is nearly always less readable than the more specific functional operations and I think we should avoid it unless we actually genuinely need its power. I think filter/map is definitely more readable and closer in structure to the original code.

File: components/ca-treeview.js [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+                    this.selected.removeAttribute('data-selected');
+                }
+
+                // select new
+                el.setAttribute('data-selected', 'true');
+
+                this.selected = el;
+
+                if (this.onclick) {
+                    const txtProp = ('textContent' in document.createElement('i')) ? 'textContent' : 'innerText';
```
Comment (tapina) [inline components/ca-treeview.js]: You're right, it's pre-IE9. One to remove asap then.

File: components/helpers/ajax.js (hunk 5 above)
Comment (tapina) [inline components/helpers/ajax.js]: s/some browsers/IE9/ ;-)

File: test/ca-tree-view-test.html (hunk 8 above)
Comment (tapina) [inline test/ca-tree-view-test.html]: What's with all the rigatoni?

File: components/ca-treeview.js (hunk 12 above)
Comment (norisalsaadie) [inline components/ca-treeview.js]: This would ban it for all projects that lint using airbnb as a base, it can be overridden here https://github.com/cambridgeweblab/eslint-config-weblab/blob/master/overrides.js, though most projects aren't using this common package, yet.

I agree... I think their bias is coming in here. But in their defence, they did label it as `good` rather than `best`, and their best examples are pretty great. So what do we do here - revert and update the eslint rules or chain array functions?

File: test/ca-tree-view-test.html (hunk 8 above)
Comment (norisalsaadie) [inline test/ca-tree-view-test.html]: This is what happens when you grew up on the internet...

File: components/helpers/ajax.js (hunk 5 above)
Comment (norisalsaadie) [inline components/helpers/ajax.js]: Okay, so I add it as a bower dependency and reflected in the examples - https://github.com/cambridgeweblab/common-ui/pull/5/commits/ccb3bbd47d7f267135d68bafec68008b674acc43

I will resolve this amd issue as part of big-red-button since I already made ajax a dependency rather than a global (but I need this merge request first to avoid conflicts and duplicate work)

File: components/ca-treeview.js (hunk 10 above)
Comment (tapina) [inline components/ca-treeview.js]: OK

File: components/ca-treeview.js [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+         * @param {object} item, schema object.
+         * @returns {HTMLElement} the newly created list item.
+         */
+        createChild(parent, tagName, item) {
+            const label = item.label || item.title || 'undefined';
+            const rel = (item.rel || '').toLowerCase().replace(/(describedby|described_by):/gi, '');
+            const attributes = {};
+
+            // get a list of attributes to be added to the element
+            // eslint-disable-next-line no-restricted-syntax
```
Comment (tapina) [inline components/ca-treeview.js]: 👍

File: components/helpers/ajax.js (hunk 5 above)
Comment (tapina) [inline components/helpers/ajax.js]: OK

File: components/ca-treeview.js [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+         */
+        get data() {
+            return this._data || [];
+        }
+
+        /**
+         * Sets the data
+         * @param {array} value, array to set as the data, defaults to empty. Invokes render.
+         */
+        set data(value = []) {
```
Comment (tapina) [inline components/ca-treeview.js]: same null problem as we had before. i didn't bother making the same comment every time i saw it...

**Comment by tapina** [comment]
I will merge
